### PR TITLE
Add obd plugin

### PR DESCRIPTION
**Pull request by Corwin-Kh**

### CODE REVIEW

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDispatcher.kt [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+							log.debug("raw_response_$responseIndex: $result")
+							dispatchResult(command, result)
+							if (command.isStale) {
+								staleCommandsCache[command] = result
+								break
+							}
+						}
+						log.info("response. ${command.name} **${response.replace('\\', '\\')}")
+					}
+				} catch (error: Throwable) {
```
Comment (alex-osm) [inline OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDispatcher.kt:107]: You should process exceptions more carefully. For example CancellationException should be handled separately

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDispatcher.kt [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+import net.osmand.shared.util.PlatformUtil
+import okio.Buffer
+import okio.IOException
+import okio.Sink
+import okio.Source
+
+
+object OBDDispatcher {
+
+	private val commandQueue = ConcurrentMutableList<OBDCommand>()
```
Comment (alex-osm) [inline OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDispatcher.kt:21]: use just List and KCollectionUtils for modifying

File: OsmAnd/src/net/osmand/plus/utils/BLEUtils.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+import android.app.Activity
+import android.bluetooth.BluetoothAdapter
+import android.bluetooth.BluetoothManager
+import android.content.Context
+import android.content.pm.PackageManager
+
+object BLEUtils {
+
+	fun isBLEEnabled(activity: Activity?): Boolean {
```
Comment (alex-osm) [inline OsmAnd/src/net/osmand/plus/utils/BLEUtils.kt:11]: Such parameter should not be nullable. 
Why isBLEEnabled returns false if activity null? If activity parameter is null - means that activity is null, but not that BLE is disabled.

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDispatcher.kt [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	private val staleCommandsCache = ConcurrentMutableMap<OBDCommand, String>()
```
[the remark this answers is not in the document]
Comment (Corwin-Kh) [reply OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDispatcher.kt]: added stale commands

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDispatcher.kt [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+										log.debug("found terminator")
+										break
+									} else {
+										log.debug("no terminator found")
+										log.debug("${resultRaw.lines().size}")
+
+									}
+								}
+							} catch (e: IOException) {
+								e.printStackTrace()
```
Comment (alex-osm) [inline OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDispatcher.kt:95]: Do not insert e.printStackTrace() anywhere in code

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDispatcher.kt [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+	fun setReadWriteStreams(readStream: Source, writeStream: Sink) {
+		scope?.cancel()
+		inputStream = readStream
+		outputStream = writeStream
+		startReadObdLooper()
+	}
+
+	fun getCommandQueue(): List<OBDCommand> {
+		return ArrayList(commandQueue)
```
Comment (alex-osm) [inline OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDispatcher.kt:157]: commandQueue is immutable, do not create duplicate list

File: OsmAnd/src/net/osmand/plus/plugins/odb/VehicleMetricsPlugin.kt [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+					val output = outputStream.sink()
+					OBDDispatcher.setReadWriteStreams(input, output)
+				}
+			}
+		} catch (error: IOException) {
+			LOG.error("Can't connect to device. $error")
+		}
+	}
+
+	fun addCommandToRead(command: OBDCommand) {
```
Comment (alex-osm) [inline OsmAnd/src/net/osmand/plus/plugins/odb/VehicleMetricsPlugin.kt:242]: There should not be references to OBDCommand outside ODB backend. This is completely backend class. There should be some metrics types or classes for operating in UI.

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDataComputer.kt [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+		}
+		if (inWindow - 1 > 0) {
+			locations = locations.subList(inWindow - 1, locations.size)
+		}
+	}
+
+	fun registerWidget(
+		type: OBDTypeWidget,
+		averageSeconds: Int,
+		formatter: OBDComputerWidgetFormatter = OBDComputerWidgetFormatter()
```
[the remark this answers is not in the document]
Comment (Corwin-Kh) [reply OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/obd/OBDDataComputer.kt:70]: A type annotation is required on a value parameter

File: OsmAnd/src/net/osmand/plus/plugins/odb/OBDTextWidget.kt [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+			COOLANT_TEMP -> {
+				obdDataWidgetType = OBDTypeWidget.TEMPERATURE_COOLANT
+				formatter = OBDComputerWidgetFormatter("%.0f")
+			}
+			FUEL_TYPE -> {
+				obdDataWidgetType = OBDTypeWidget.FUEL_TYPE
+				formatter = OBDFuelTypeFormatter()
+			}
+		}
+		widgetComputer = OBDDataComputer.registerWidget(obdDataWidgetType, 15, formatter)
```
[the remark this answers is not in the document]
Comment (Corwin-Kh) [reply OsmAnd/src/net/osmand/plus/plugins/odb/OBDTextWidget.kt:68]: This is temporary value, should be changed when implementing UI for widgets. Added todo here

**Pull request merged**